tests/RobotFramework/tests/tedge_agent/workflows/custom_restart.toml

```diff
@@ -0,0 +1,23 @@
+operation = "controled_restart"
+
+[init]
+script = "/etc/tedge/operations/log-restart.sh ${.topic.cmd_id} ${.payload.status}"
+next = ["scheduled"]
+
+[scheduled]
+script = "restart"
```

Comment:
As you highlighted already, I'd also be in favour of introducing a different `action` key instead of overloading the `script` key.

Reply:
The proposal is to with something along these lines:

```
trigger = {
    action = "restart",
    on_exec = "restarting",
    on_success = "successful_restart",
    on_error = "failed_restart"
}
```

Reply:
Why not continue using the `next` field itself for those successful and failed target states instead of this new `on_success` and `on_error` directives? So that it is consistent with the rest of the workflow actions?

That `on_exec` state is still a bit confusing to me. A user looking at the workflow definition to make sense out of the end-to-end control flow will be a bit lost seeing that `restarting` state, as there is no other reference to that state in the workflow file. I guess I'm still missing something about this contract.

Reply:
As discussed offline, that `restarting` state being implicit, while the `successful_restart` and `failed_restart` states being explicit was the confusing aspect for me. One way to make things clearer would be to have that `restarting` state also explicitly defined in the workflow as follows:

```toml
operation = "controlled_restart"

# ...

[scheduled]
action = "restart"
next = ["restarting"]

[restarting]
action = "check-restart"
next = ["successful_restart", "failed_restart"]

[successful_restart]
script = "/etc/tedge/operations/log-restart.sh ${.topic.cmd_id} ${.payload.status}"
next = ["successful"]

[failed_restart]
script = "/etc/tedge/operations/log-restart.sh ${.topic.cmd_id} ${.payload.status} ${.payload.restartError}"
next = ["failed"]

# ...
```

The issue with this approach is that we're forcing the user to explicitly define additional states as per the internal contract of the `restart` action. The need to have an explicit `restarting` state may not be that intuitive to a user.

Another option is to completely hide the `restarting` state by allowing the agent to auto-generate its own internal `restarting` states without exposing it to the user in the workflow files. For example, if the `restart` action is called from the `scheduled` state, then the `status` will be updated form `scheduled` to `scheduled_restarting`. If another state `xyz` calls this action, then it will be moved to `xyz_restarting` state. That way, we can avoid state value conflicts as well. The only issue with this approach is that the user will notice these "internal states" that he did not define in his workflow file. But, a minor issue IMO. The user should be expecting it while relying complex inbuilt actions like `restart`.